File: src/board.py
import random
import pygame
from itertools import combinations

from card import Card

# RCade game dimensions
WIDTH = 336
HEIGHT = 262
FPS = 60

# Colors
CARD_BG = (240, 240, 235)
CARD_BG_HOVER = (200, 200, 195)
CARD_BG_SELECTED = (255, 255, 250)
CARD_BORDER = (60, 60, 60)
SELECTED_BORDER = (255, 215, 0)
CURSOR_COLOR = (100, 200, 255)
FLASH_GREEN = (50, 205, 50)
FLASH_RED = (220, 50, 50)

# Categories
CATEGORY_1 = ["circle", "square", "triangle"]  # shape
CATEGORY_2 = ["red", "green", "blue"]  # color
CATEGORY_3 = ["1", "2", "3"]  # count
CATEGORY_4 = ["solid", "striped", "open"]  # fill

# Grid layout
CARD_WIDTH = 90
CARD_HEIGHT = 60
CARD_MARGIN = 10
GRID_COLS = 3
GRID_ROWS = 3
GRID_START_X = (WIDTH - (CARD_WIDTH * GRID_COLS + CARD_MARGIN * (GRID_COLS - 1))) // 2
GRID_START_Y = (HEIGHT - (CARD_HEIGHT * GRID_ROWS + CARD_MARGIN * (GRID_ROWS - 1))) // 2


class GameBoard:
    def __init__(self):
        self.deck = self._create_deck()
        random.shuffle(self.deck)
        self.cards = [[None for _ in range(GRID_COLS)] for _ in range(GRID_ROWS)]
        self.selected = set()  # Set of (row, col) tuples
        self.cursor_row = 0
        self.cursor_col = 0
        self._deal_initial_cards()
        self._prev_inputs = {"up": False, "down": False, "left": False, "right": False, "a": False}
        self.score = 0
        self.flash_color = None  # None, FLASH_GREEN, or FLASH_RED
        self.flash_timer = 0
        self.flash_duration = FPS // 2  # Half second
        self.flash_positions = set()  # Positions to flash
        self.game_over = False  # True when no valid sets and deck is empty
        self.debug_mode = True
        self._print_valid_sets()

    def _print_valid_sets(self):
        """Print all valid sets to console (debug mode only)."""
        if not self.debug_mode:
            return
        valid_sets = self._get_all_valid_sets()
        print(f"\n=== Valid sets ({len(valid_sets) // 6}): ===")  # Divide by 6 to account for permutations
        seen = set()
        for s in valid_sets:
            key = tuple(sorted(s))
            if key not in seen:
                seen.add(key)
                print(f"  {s[0]} + {s[1]} + {s[2]}")

    def _get_all_valid_sets(self):
        """Get all valid sets on the board."""
        card_positions = [(row, col) for row in range(GRID_ROWS) for col in range(GRID_COLS)
                         if self.cards[row][col] is not None]
        valid_sets = []
        for pos1, pos2, pos3 in combinations(card_positions, 3):
            if self.cards[pos1[0]][pos1[1]].check_if_set(
                self.cards[pos2[0]][pos2[1]], self.cards[pos3[0]][pos3[1]]):
                valid_sets.append((pos1, pos2, pos3))
        return valid_sets

    def _has_valid_set(self):
        """Quick check if any valid set exists on the board."""
        card_positions = [(row, col) for row in range(GRID_ROWS) for col in range(GRID_COLS)
                         if self.cards[row][col] is not None]
        for pos1, pos2, pos3 in combinations(card_positions, 3):
            if self.cards[pos1[0]][pos1[1]].check_if_set(
                self.cards[pos2[0]][pos2[1]], self.cards[pos3[0]][pos3[1]]):
                return True
        return False

    def _create_deck(self):
        """Create the full 81-card deck."""
        deck = []
        for c1 in CATEGORY_1:
            for c2 in CATEGORY_2:
                for c3 in CATEGORY_3:
                    for c4 in CATEGORY_4:
                        deck.append(Card(c1, c2, c3, c4))
        return deck

    def _deal_initial_cards(self):
        """Deal 9 cards to fill the board, ensuring at least one valid set exists."""
        for row in range(GRID_ROWS):
            for col in range(GRID_COLS):
                if self.deck:
                    self.cards[row][col] = self.deck.pop()
        self._ensure_valid_set_exists()

    def _ensure_valid_set_exists(self):
        """Swap cards with deck until a valid set exists on the board."""
        max_attempts = 1000
        attempts = 0
        while not self._has_valid_set() and self.deck and attempts < max_attempts:
            # Get all board positions with cards
            board_positions = [(r, c) for r in range(GRID_ROWS) for c in range(GRID_COLS)
                              if self.cards[r][c] is not None]
            if not board_positions:
                break
            # Swap a random board card with a random deck card
            r, c = random.choice(board_positions)
            deck_idx = random.randrange(len(self.deck))
            self.cards[r][c], self.deck[deck_idx] = self.deck[deck_idx], self.cards[r][c]
            attempts += 1

    def _replace_cards(self, positions):
        """Replace cards at the given positions with new ones from the deck."""
        for r, c in positions:
            if self.deck:
                self.cards[r][c] = self.deck.pop()
            else:
                self.cards[r][c] = None

    def get_card_rect(self, row, col):
        """Get the rectangle for a card at the given grid position."""
        x = GRID_START_X + col * (CARD_WIDTH + CARD_MARGIN)
        y = GRID_START_Y + row * (CARD_HEIGHT + CARD_MARGIN)
        return pygame.Rect(x, y, CARD_WIDTH, CARD_HEIGHT)

    def update(self, inputs):
        """Handle input for cursor movement and card selection."""
        # Update flash timer
        if self.flash_timer > 0:
            self.flash_timer -= 1
            if self.flash_timer == 0:
                self.flash_color = None
                self.flash_positions.clear()

        # Cursor movement (edge-triggered)
        if inputs["up"] and not self._prev_inputs["up"]:
            self.cursor_row = (self.cursor_row - 1) % GRID_ROWS
        if inputs["down"] and not self._prev_inputs["down"]:
            self.cursor_row = (self.cursor_row + 1) % GRID_ROWS
        if inputs["left"] and not self._prev_inputs["left"]:
            self.cursor_col = (self.cursor_col - 1) % GRID_COLS
        if inputs["right"] and not self._prev_inputs["right"]:
            self.cursor_col = (self.cursor_col + 1) % GRID_COLS

        # Card selection with A button
        if inputs["a"] and not self._prev_inputs["a"]:
            pos = (self.cursor_row, self.cursor_col)
            if pos in self.selected:
                self.selected.remove(pos)
            elif len(self.selected) < 3:
                self.selected.add(pos)
                # Check if we have exactly 3 cards selected
                if len(self.selected) == 3:
                    self._check_set()


        self._prev_inputs = {k: inputs[k] for k in self._prev_inputs}

    def _check_set(self):
        """Check if the 3 selected cards form a valid set."""
        positions = list(self.selected)
        cards = [self.cards[r][c] for r, c in positions]

        # Store positions for flashing before clearing selected
        self.flash_positions = set(positions)
        self.flash_timer = self.flash_duration

        if cards[0].check_if_set(cards[1], cards[2]):
            # Valid set
            self.flash_color = FLASH_GREEN
            self.score += 1
            # Replace cards with new ones from deck
            for r, c in positions:
                if self.deck:
                    self.cards[r][c] = self.deck.pop()
                else:
                    self.cards[r][c] = None
            self._ensure_valid_set_exists()
            self._print_valid_sets()
            # Check for game over: no valid sets and deck is empty
            if not self.deck and not self._has_valid_set():
                self.game_over = True
        else:
            # Invalid set
            self.flash_color = FLASH_RED

        self.selected.clear()

    def draw(self, screen):
        """Draw the game board with all cards."""
        for row in range(GRID_ROWS):
            for col in range(GRID_COLS):
                card = self.cards[row][col]
                if card is None:
                    continue

                rect = self.get_card_rect(row, col)
                is_selected = (row, col) in self.selected
                is_cursor = (row == self.cursor_row and col == self.cursor_col)

                # Card background - darker for hover (including when selected), lighter for selected only
                if is_cursor:
                    pygame.draw.rect(screen, CARD_BG_HOVER, rect)
                elif is_selected:
                    pygame.draw.rect(screen, CARD_BG_SELECTED, rect)
                else:
                    pygame.draw.rect(screen, CARD_BG, rect)

                # Border: flash color overrides, then selected = gold, cursor = blue, normal = dark
                is_flashing = (row, col) in self.flash_positions
                if self.flash_color and is_flashing:
                    pygame.draw.rect(screen, self.flash_color, rect, 3)
                elif is_selected:
                    pygame.draw.rect(screen, SELECTED_BORDER, rect, 3)
                elif is_cursor:
                    pygame.draw.rect(screen, CURSOR_COLOR, rect, 2)
                else:
                    pygame.draw.rect(screen, CARD_BORDER, rect, 1)

                # Draw card shapes
                card.draw(screen, rect.x, rect.y, rect.width, rect.height)

File: src/board.ts
import type { RGB, PlayerInput } from './types';
import { Card, createDeck, shuffleDeck, hasValidSet, findAllValidSets } from './card';
import { Renderer } from './renderer';
import {
  GRID_COLS, GRID_ROWS, GRID_START_X, GRID_START_Y,
  CARD_WIDTH, CARD_HEIGHT, CARD_MARGIN,
  CARD_BG, CARD_BG_HOVER, CARD_BG_SELECTED, CARD_BORDER,
  SELECTED_BORDER, CURSOR_COLOR, FLASH_GREEN, FLASH_RED,
  FLASH_DURATION_MS, FLIP_DURATION_MS,
} from './constants';

interface FlipAnimation {
  oldCard: Card | null;
  newCard: Card | null;
  startTime: number;
}

export class GameBoard {
  deck: Card[];
  cards: (Card | null)[][];
  selected: Set<string>;
  cursorRow: number;
  cursorCol: number;
  score: number;
  gameOver: boolean;
  debugMode: boolean;

  private prevInputs: Record<string, boolean>;
  private flashColor: RGB | null;
  private flashStartTime: number;
  private flashPositions: Set<string>;
  private flipAnimations: Map<string, FlipAnimation>;

  constructor(initialInputs?: PlayerInput) {
    this.deck = createDeck();
    shuffleDeck(this.deck);
    this.cards = Array.from({ length: GRID_ROWS }, () => Array(GRID_COLS).fill(null));
    this.selected = new Set();
    this.cursorRow = 0;
    this.cursorCol = 0;
    this.score = 0;
    this.gameOver = false;
    this.debugMode = true;

    this.prevInputs = {
      up: initialInputs?.up ?? false,
      down: initialInputs?.down ?? false,
      left: initialInputs?.left ?? false,
      right: initialInputs?.right ?? false,
      a: initialInputs?.a ?? false,
    };

    this.flashColor = null;
    this.flashStartTime = 0;
    this.flashPositions = new Set();
    this.flipAnimations = new Map();

    this.dealInitialCards();
    this.printValidSets();
  }

  private dealInitialCards(): void {
    for (let row = 0; row < GRID_ROWS; row++) {
      for (let col = 0; col < GRID_COLS; col++) {
        if (this.deck.length > 0) {
          this.cards[row][col] = this.deck.pop()!;
        }
      }
    }
    this.ensureValidSetExists();
  }

  private ensureValidSetExists(): void {
    const flatCards = this.cards.flat();
    let attempts = 0;
    const maxAttempts = 1000;

    while (!hasValidSet(flatCards) && this.deck.length > 0 && attempts < maxAttempts) {
      const boardPositions: [number, number][] = [];
      for (let r = 0; r < GRID_ROWS; r++) {
        for (let c = 0; c < GRID_COLS; c++) {
          if (this.cards[r][c]) boardPositions.push([r, c]);
        }
      }
      if (boardPositions.length === 0) break;

      const [r, c] = boardPositions[Math.floor(Math.random() * boardPositions.length)];
      const deckIdx = Math.floor(Math.random() * this.deck.length);
      const temp = this.cards[r][c];
      this.cards[r][c] = this.deck[deckIdx];
      this.deck[deckIdx] = temp!;
      attempts++;
    }
  }

  private printValidSets(): void {
    if (!this.debugMode) return;
    const validSets = findAllValidSets(this.cards.flat());
    console.log(`\n=== Valid sets (${validSets.length}): ===`);
    for (const [i, j, k] of validSets) {
      const r1 = Math.floor(i / GRID_COLS), c1 = i % GRID_COLS;
      const r2 = Math.floor(j / GRID_COLS), c2 = j % GRID_COLS;
      const r3 = Math.floor(k / GRID_COLS), c3 = k % GRID_COLS;
      console.log(`  (${r1},${c1}) + (${r2},${c2}) + (${r3},${c3})`);
    }
  }

  private posKey(row: number, col: number): string {
    return `${row},${col}`;
  }

  getCardRect(row: number, col: number): { x: number; y: number; width: number; height: number } {
    return {
      x: GRID_START_X + col * (CARD_WIDTH + CARD_MARGIN),
      y: GRID_START_Y + row * (CARD_HEIGHT + CARD_MARGIN),
      width: CARD_WIDTH,
      height: CARD_HEIGHT,
    };
  }

  update(inputs: PlayerInput, now: number): void {
    // Update flash
    if (this.flashColor && now - this.flashStartTime > FLASH_DURATION_MS) {
      this.flashColor = null;
      this.flashPositions.clear();
    }

    // Update flip animations
    for (const [key, anim] of this.flipAnimations) {
      if (now - anim.startTime > FLIP_DURATION_MS) {
        this.flipAnimations.delete(key);
      }
    }

    // Edge-triggered cursor movement
    if (inputs.up && !this.prevInputs.up) {
      this.cursorRow = (this.cursorRow - 1 + GRID_ROWS) % GRID_ROWS;
    }
    if (inputs.down && !this.prevInputs.down) {
      this.cursorRow = (this.cursorRow + 1) % GRID_ROWS;
    }
    if (inputs.left && !this.prevInputs.left) {
      this.cursorCol = (this.cursorCol - 1 + GRID_COLS) % GRID_COLS;
    }
    if (inputs.right && !this.prevInputs.right) {
      this.cursorCol = (this.cursorCol + 1) % GRID_COLS;
    }

    // Card selection with A button
    if (inputs.a && !this.prevInputs.a) {
      const key = this.posKey(this.cursorRow, this.cursorCol);
      if (this.selected.has(key)) {
        this.selected.delete(key);
      } else if (this.selected.size < 3 && this.cards[this.cursorRow][this.cursorCol]) {
        this.selected.add(key);
        if (this.selected.size === 3) {
          this.checkSet(now);
        }
      }
    }

    this.prevInputs = {
      up: inputs.up,
      down: inputs.down,
      left: inputs.left,
      right: inputs.right,
      a: inputs.a,
    };
  }

  private checkSet(now: number): void {
    const positions = Array.from(this.selected).map(k => {
      const [r, c] = k.split(',').map(Number);
      return { row: r, col: c };
    });

    this.flashPositions = new Set(this.selected);
    this.flashStartTime = now;

    const cards = positions.map(p => this.cards[p.row][p.col]!);

    if (cards[0].checkIfSet(cards[1], cards[2])) {
      this.flashColor = FLASH_GREEN;
      this.score++;

      // Start flip animations and replace cards
      for (const pos of positions) {
        const key = this.posKey(pos.row, pos.col);
        const oldCard = this.cards[pos.row][pos.col];
        const newCard = this.deck.length > 0 ? this.deck.pop()! : null;

        this.flipAnimations.set(key, {
          oldCard,
          newCard,
          startTime: now,
        });
        this.cards[pos.row][pos.col] = newCard;
      }

      this.ensureValidSetExists();

      // Update animations with potentially swapped cards
      for (const pos of positions) {
        const key = this.posKey(pos.row, pos.col);
        const anim = this.flipAnimations.get(key);
        if (anim) {
          anim.newCard = this.cards[pos.row][pos.col];
        }
      }

      this.printValidSets();

      // Check for game over
      if (this.deck.length === 0 && !hasValidSet(this.cards.flat())) {
        this.gameOver = true;
      }
    } else {
      this.flashColor = FLASH_RED;
    }

    this.selected.clear();
  }

  draw(renderer: Renderer, now: number): void {
    const ctx = renderer.getContext();

    for (let row = 0; row < GRID_ROWS; row++) {
      for (let col = 0; col < GRID_COLS; col++) {
        const key = this.posKey(row, col);
        const baseRect = this.getCardRect(row, col);
        let card = this.cards[row][col];
        let rect = baseRect;

        // Handle flip animation
        const anim = this.flipAnimations.get(key);
        if (anim) {
          const elapsed = now - anim.startTime;
          const progress = Math.min(1, elapsed / FLIP_DURATION_MS);

          let scale: number;
          if (progress < 0.5) {
            card = anim.oldCard;
            scale = 1 - progress * 2;
          } else {
            card = anim.newCard;
            scale = (progress - 0.5) * 2;
          }

          if (!card) continue;

          const scaledWidth = Math.floor(CARD_WIDTH * scale);
          if (scaledWidth < 2) continue;

          const xOffset = Math.floor((CARD_WIDTH - scaledWidth) / 2);
          rect = {
            x: baseRect.x + xOffset,
            y: baseRect.y,
            width: scaledWidth,
            height: CARD_HEIGHT,
          };
        }

        if (!card) continue;

        const isSelected = this.selected.has(key);
        const isCursor = row === this.cursorRow && col === this.cursorCol;

        // Card background
        let bgColor: RGB = CARD_BG;
        if (isCursor) bgColor = CARD_BG_HOVER;
        else if (isSelected) bgColor = CARD_BG_SELECTED;
        renderer.fillRect(rect.x, rect.y, rect.width, rect.height, bgColor);

        // Border
        const isFlashing = this.flashPositions.has(key);
        let borderColor: RGB = CARD_BORDER;
        let borderWidth = 1;

        if (this.flashColor && isFlashing) {
          borderColor = this.flashColor;
          borderWidth = 3;
        } else if (isSelected) {
          borderColor = SELECTED_BORDER;
          borderWidth = 3;
        } else if (isCursor) {
          borderColor = CURSOR_COLOR;
          borderWidth = 2;
        }

        renderer.strokeRect(rect.x, rect.y, rect.width, rect.height, borderColor, borderWidth);

        // Draw card shapes
        card.draw(ctx, rect.x, rect.y, rect.width, rect.height);
      }
    }
  }
}
